### DISCUSSION TITLE
Confusion about Result.zip extensions

### DISCUSSION THREAD

**Discussion by frko** [arrow-kt/arrow#2802] [Q&A] [answered]
I have been looking at https://github.com/arrow-kt/arrow/blob/5d0bc9e214f9b92f8492a8202a7c165d5e81fa9e/arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Result.kt#L181 for a while and I am slightly confused about its implementation.

I wonder if it is intentional that the implementation will throw an NPE when a Result<T> contains null as success value, or, at least I think it does looking at the source.

Regards, Frank

**Comment by nomisRev** [comment]
Hey Frank,

Okay, I see what you mean, and this is definitely a bug.

``` kotlin
val r = Result.success<Int?>(null)
r.zip(r) { x, y -> (x?.plus(y ?: 0)) ?: -1 } // shouldBe -1 but is java.lang.NullPointerException
```

The proper implementation would check for `isSuccess` and not rely on `null` for `success`, similar to how `Result` works internally.

```kotlin
inline fun <A, B, C> Result<A>.zip(b: Result<B>, transform: (A, B) -> C): Result<C> =
  if (isSuccess && b.isSuccess) Result.success(transform(getOrNull() as A, b.getOrNull() as B))
  else Result.failure((exceptionOrNull()?.apply { addSuppressed(b.exceptionOrNull()) } ?: b.exceptionOrNull())!!)
```

**Reply by frko** [reply]
Great you looked into this that fast .

Was wondering about the usage of suppressed  for exception collection here as it feels off to me semantically. Is this to prevent 'yet another exception' to be created with it's own stacktrace while in fact the situation it is handling is not exceptional?

**Reply by nomisRev** [reply]
Hey @frko,

It's to prevent a stack-trace from getting lost. Let's take following example:

```kotlin
val runtime = Result.failure<Int>(RuntimeException("BOOM!"))
val state = Result.failure<Int>(IllegalStateException("Wrong Builder State!"))

fun main() {
  val res = runtime.zip(state) { x, y -> x + y }.getOrThrow()
  println(res)
}
```

Without `addSuppressed` of _state_ into _runtime_ we will get:

```text
Exception in thread "main" java.lang.RuntimeException: BOOM!
	at arrow.core.continuations.MainKt.<clinit>(Main.kt:8)
```

If we add the suppression:

```text
Exception in thread "main" java.lang.RuntimeException: BOOM!
	at arrow.core.continuations.MainKt.<clinit>(Main.kt:8)
	Suppressed: java.lang.IllegalStateException: Wrong Builder State!
		at arrow.core.continuations.MainKt.<clinit>(Main.kt:9)
```

So through `addSuppressed` we make sure that no information is lost when combining different `Result`.

**Comment by nomisRev** [comment, marked as answer]
I opened a PR with a fix https://github.com/arrow-kt/arrow/pull/2804
